### PR TITLE
Make Android respect power saving/disabled animations

### LINKED ISSUE #16476: FadeTo not completing on Samsung devices with visibility enhancement "Remove animations" turned on
### Description

When turning on accessibility feature "Remove animations" on Samsung devices, FadeTo is not completing.

Settings > Accessibility > Visibility enhancements > Remove animations = turn on

This issue exists in Xamarin.Forms as well and seems to be a regression (fixed in Xamarin.Forms 4.4.0):
https://github.com/xamarin/Xamarin.Forms/issues/5686
https://github.com/xamarin/Xamarin.Forms/issues/15743

### Steps to Reproduce

1. Create new .NET MAUI app project
2. add `await CounterBtn.FadeTo(0.5);` in `private async void OnCounterClicked(object sender, EventArgs e)`
3. turn on Settings > Accessibility > Visibility enhancements > Remove animations (Samsung device required)
4. run and tap "Click me" button
5. CounterBtn.Text is not updating

When tapping "Click me" again, the first called `await CounterBtn.FadeTo(0.5);` completes and the text updates. I guess this happens as FadeTo is completing as soon as another FadeTo on the same button is started.

Besides this the opacity of the button is not changed (should be 0.5).

### Link to public reproduction project repository

https://github.com/eichbaum/FadeToDemo

### Version with bug

7.0.86

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android 13, One UI Core Version 5.1

### Did you find any workaround?

turn Settings > Accessibility > Visibility enhancements > Remove animations off

### Relevant log output

_No response_

### LINKED ISSUE #16478: Animation CallBack action not invoked when enabled the battery-saver in Android device.
### Description

I encountered an issue with the custom animation control developed to replace the framework's ActivityIndicator. The problem arises when the device's battery-saver mode is enabled. Under this condition, the Animation CallBack Action, which is a crucial part of the animation functionality, fails to get invoked.

We're seeking a solution that guarantees the consistent triggering of the Animation CallBack Action, irrespective of the device's battery saver mode being on or off. This would ensure seamless and continuous animations even in battery-saving scenarios, thereby improving the overall user experience.battery-saver

https://github.com/dotnet/maui/assets/97146406/4769f439-b48a-4b76-a13d-1a8f8037dd03

### Steps to Reproduce

1. Launch the .NET MAUI application on an Android device running Android OS version X or higher.
2. Ensure the application includes a label and a button.
3. Enable the device's battery saver mode either through the Settings menu or by using the battery status notification.
4. Tap the button to trigger the custom animation, which should scale the label.

### Link to public reproduction project repository

https://github.com/anandhan-rajagopal/CustomAnimation

### Version with bug

7.0.49

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android 5.0 and above

### Did you find any workaround?

https://github.com/dotnet/maui/blob/main/src/Core/src/Animations/AnimationManager.cs#L38,
https://github.com/dotnet/maui/blob/main/src/Core/src/Animations/PlatformTicker.Android.cs#L61

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by hartez**
When the animation methods and tickers were ported from Forms to MAUI, the bits which pay attention to the Android system animation settings did not make the port. These changes bring those bits forward:

- If the system's animations are disabled (either by accessibility features or developer features), new animations will jump immediately to their finished state

- If the device's power saving mode kicks on while animations are in progress, the animations will immediately jump to their finished state

- If the device's animation durations is set to zero (disabled) while animations are in progress and the API version is 33 or greater, the animations will immediately jump to their finished state

Fixes #16476. 
Fixes #16478.

This should make the tests from #16511 pass.

### CODE REVIEW

**Review by PureWeen — CHANGES_REQUESTED** [review]
DisablingTickerFinishesAllAnimationsInChain failing on mac

**Review by StephaneDelcroix — DISMISSED** [review]

File: src/Core/src/PublicAPI/net-tizen/PublicAPI.Shipped.txt [hunk 1]
```diff
@@ -2919,3 +2919,5 @@ virtual Microsoft.Maui.WindowOverlay.HandleUIChange() -> void
 virtual Microsoft.Maui.WindowOverlay.Initialize() -> bool
 virtual Microsoft.Maui.WindowOverlay.RemoveWindowElement(Microsoft.Maui.IWindowOverlayElement! drawable) -> bool
 virtual Microsoft.Maui.WindowOverlay.RemoveWindowElements() -> void
+virtual Microsoft.Maui.Animations.Ticker.OnSystemEnabledChanged() -> void
```
Comment (PureWeen) [inline src/Core/src/PublicAPI/net-tizen/PublicAPI.Shipped.txt:2922]: Should this be moved to UnShipped?

**Review by PureWeen — DISMISSED** [review]

**Comment by hartez** [comment]
Converting to draft until I can figure out why `DisablingTickerPreventsAnimationFromRepeating` fails intermittently.

File: src/Controls/src/Core/Tweener.cs [hunk 2]
```diff
@@ -39,27 +38,40 @@ public TweenerAnimation(Func<long, bool> step)
...
+			if (HasFinished)
+			{
+				return;
+			}
+
+			HasFinished = true;
+
+			// The tweeners use long.MaxValue for in-band signaling that they should
+			// jump to the end of the animation 
+			_ = _step.Invoke(long.MaxValue);
```
Comment (mattleibow) [inline src/Controls/src/Core/Tweener.cs:59]: Should this be a FireAndForget case? Exceptions may be swallowed.

File: src/Controls/src/Core/AnimationExtensions.cs [hunk 3]
```diff
@@ -36,7 +37,14 @@ namespace Microsoft.Maui.Controls
...
+		// We use a ConcurrentDictionary because Tweener relies on being able to remove
+		// animations from the AnimationManager within its finalizer (via the Remove extension
+		// method below). Since finalization occurs on a different thread, it risks crashes when
+		// the finalizer is running at the same time another animation is finsihing and removing
+		// itself from this dictionary. So until we can change that design, this dictionary must
+		// be thread-safe.
+		static readonly ConcurrentDictionary<int, Animation> s_tweeners;
+		
 		static readonly Dictionary<AnimatableKey, Info> s_animations;
 		static readonly Dictionary<AnimatableKey, int> s_kinetics;
```
Comment (mattleibow) [inline src/Controls/src/Core/AnimationExtensions.cs:49]: These 2 dictionaries are modified at the same time as the tweeners, do they also need to be concurrent?

File: src/Controls/src/Core/Tweener.cs [hunk 4]
```diff
@@ -69,119 +81,127 @@ public Tweener(uint length, uint rate, IAnimationManager animationManager)
...
+			_animationManager = animationManager;
 			Loop = false;
 		}
 
 		public AnimatableKey Handle { get; set; }
 
 		public uint Length { get; }
 
-		public uint Rate { get; }
+		public uint Rate { get; } = 1;
```
Comment (mattleibow) [inline src/Controls/src/Core/Tweener.cs:92]: Why is this needed? It is set in the ctor.

File: src/Controls/tests/Core.UnitTests/TestClasses/AnimationReadyHandler.cs [hunk 5]
```diff
@@ -83,7 +89,12 @@ public object GetService(Type serviceType)
 				if (serviceType == typeof(IAnimationManager))
 					return _animationManager;
 
-				throw new NotSupportedException();
+				if (serviceType == typeof(IDispatcherProvider))
+				{
+					return DispatcherProvider.Current;
+				}
```
Comment (mattleibow) [inline src/Controls/tests/Core.UnitTests/TestClasses/AnimationReadyHandler.cs:95]: What was this provider used for in the new code?

File: src/Core/src/PublicAPI/net-android/PublicAPI.Shipped.txt [hunk 6]
```diff
@@ -2068,7 +2068,7 @@ override Microsoft.Maui.Animations.LerpingAnimation.Update(double percent) -> vo
 override Microsoft.Maui.Animations.PlatformTicker.IsRunning.get -> bool
 override Microsoft.Maui.Animations.PlatformTicker.Start() -> void
 override Microsoft.Maui.Animations.PlatformTicker.Stop() -> void
-override Microsoft.Maui.Animations.PlatformTicker.SystemEnabled.get -> bool
+
```
Comment (mattleibow) [inline src/Core/src/PublicAPI/net-android/PublicAPI.Shipped.txt:2071]: For the tools to make a nice diff, this has to be done via removing it in the Unshipped.

File: src/Controls/src/Core/AnimationExtensions.cs (hunk 3 above)
Comment (hartez) [reply src/Controls/src/Core/AnimationExtensions.cs:49]: No. They aren't modified from a finalizer. They're only modified on the UI thread by definition.

File: src/Controls/tests/Core.UnitTests/TestClasses/AnimationReadyHandler.cs (hunk 5 above)
Comment (hartez) [reply src/Controls/tests/Core.UnitTests/TestClasses/AnimationReadyHandler.cs:95]: Before I wrote all the infrastructure to force the tests to run on a single thread, it was added so the tests wouldn't throw a NotSupportedException when something needed a dispatcher. It's probably not be necessary anymore.

File: src/Controls/src/Core/Tweener.cs (hunk 4 above)
Comment (hartez) [reply src/Controls/src/Core/Tweener.cs:91]: Okay, removed from constructor.

**Review by mattleibow — APPROVED** [review]
